File: BackEnd-Siukat/services/keringanan.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
	"time"
)

type KeringananService struct{}

func (s *KeringananService) Add(req models.Keringanan, atribut string) (models.Keringanan, error) {
	req.Atribut = atribut
	err := config.DB.Create(&req).Error
	return req, err
}

func (s *KeringananService) Edit(req models.Keringanan, noPeserta string, existing models.Keringanan, atribut string) (models.Keringanan, error) {
	data := map[string]interface{}{}
	data["flag"] = req.Flag

	if req.Flag == "" {
		data["scan_keringanan"] = ""
	} else if req.ScanKeringanan != "" {
		data["scan_keringanan"] = req.ScanKeringanan
	} else {
		data["scan_keringanan"] = existing.ScanKeringanan
	}

	if err := config.DB.Model(&models.Keringanan{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(data).Error; err != nil {
		return models.Keringanan{}, err
	}
	var res models.Keringanan
	config.DB.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&res)
	return res, nil
}

func (s *KeringananService) AddLog(user models.Keringanan, atribut string, executor string, timestamp *time.Time) error {
	logLst := models.LogKeringanan{
		NoPeserta:      user.NoPeserta,
		ScanKeringanan: user.ScanKeringanan,
		Flag:           user.Flag,
		Atribut:        atribut,
		Executor:       executor,
		Timestamp:      timestamp,
	}
	return config.DB.Create(&logLst).Error
}

func (s *KeringananService) GetByLoggedIn(noPeserta string) (models.Keringanan, error) {
	var res models.Keringanan
	err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error
	return res, err
}

func (s *KeringananService) CheckData(noPeserta string) (bool, error) {
	var res models.Keringanan
	if err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error; err != nil {
		return false, err
	}

	// Any required logic
	return true, nil
}

File: Be-siukat-new/siukat-service-dev_snmptn_2022/services/keringanan.js
var Keringanan = require('../models/keringanan');
var LogKeringanan = require('../models/log_keringanan');

var keringanan = function () { };

keringanan.prototype.add = function (req, atribut) {
    return new Promise(function (resolve, reject) {
        Keringanan.create({
            no_peserta: req.no_peserta,
            scan_keringanan: req.scan_keringanan,
            flag: req.flag,
            atribut: atribut
        }, {
            fields: [
                'no_peserta',
                'scan_keringanan',
                'flag',
                'atribut'
            ]
        }).then(function (response) {
            resolve(response)
        }).catch(function (err) {
            reject(err)
        })
    })
}

keringanan.prototype.edit = function (req, no_peserta, users, atribut) {
    return new Promise(function (resolve, reject) {
        var data = {
            scan_keringanan: users.scan_keringanan,
            flag: req.body.flag,
        }

        if (req.file !== undefined) {
            data.scan_keringanan = req.body.flag == "" ? "" : req.file.filename;
        }

        Keringanan.update(data, {
            where: {
                no_peserta: no_peserta,
                atribut: atribut
            }
        }).then(function (response) {
            resolve(response)
        }).catch(function (err) {
            reject(err)
        })
    })
}

keringanan.prototype.addLog = function (users, atribut, executor, timestamp) {
    return new Promise(function (resolve, reject) {
        LogKeringanan.create({
            no_peserta: users.no_peserta,
            scan_keringanan: users.scan_keringanan,
            flag: users.flag,
            atribut: atribut,
            executor: executor,
            timestamp: timestamp
        }, {
            field: [
                'no_peserta',
                'scan_keringanan',
                'flag',
                'atribut',
                'executor',
                'atribut'
            ]
        }).then(function (response) {
            resolve(response)
        }).catch(function (err) {
            reject(err)
        })
    })
}

keringanan.prototype.getByLoggedIn = function (no_peserta) {
    return new Promise(function (resolve, reject) {
        Keringanan.findOne({
            where: {
                no_peserta: no_peserta
            }
        }).then(function (response) {
            resolve(response)
        }).catch(function (err) {
            reject(err)
        })
    })
}

keringanan.prototype.checkData = function (no_peserta) {
    return new Promise((resolve, reject) => {
        Keringanan.findOne({
            where: {
                no_peserta: no_peserta
            },
            raw: true,
        }).then(function (response) {
            var result = true

            for (var key in response) {
                result *= (response[key] !== 0 && response[key] !== "")
            }
            resolve(result)
        }).catch(function (err) {
            reject(err)
        })
    })
}

module.exports = keringanan
